---
title: "Single-cell RNA-seq Preprocessing"
output:
  html_notebook: default
  html_document: default
  pdf_document: default
---


# Libraries & Functions

Load libraries and some functions to be used on all 4 datasets.

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(gridExtra)
library(optparse)
library(cowplot)
library(knitr)
```


```{r}
load_dataset = function(file, name) {
    print(paste0('Loading dataset ', file))
    gene_expression_matrix = read.csv(file, row.names=1)  # Load csv file
    seuratObj = CreateSeuratObject(counts=gene_expression_matrix, project=name)  # Create Seurat object
    print(cat(name, dim(seuratObj)))
    return(seuratObj)
}
```

```{r}
quality_control_plots = function(data, name) {
    # Add percentage of mitochondrial genes to the Seurat object
    data[['percent.mt']] = PercentageFeatureSet(data, pattern = '^mt-')

    # Visualize QC metrics as a violin plot
    violin_plot = VlnPlot(data, features = c('nFeature_RNA', 'nCount_RNA', 'percent.mt'), ncol = 3)
    # Visualize relationship between nCount_RNA and nFeature_RNA
    scatter_plot_1 = FeatureScatter(data, feature1 = 'nCount_RNA', feature2 = 'percent.mt')
    scatter_plot_2 = FeatureScatter(data, feature1 = 'nCount_RNA', feature2 = 'nFeature_RNA')
    scatter_plot = scatter_plot_1 + scatter_plot_2
    plot(violin_plot)
    plot(scatter_plot)
    return(data)
}
```


```{r}
run_plot_PCA = function(data) {
    data_pca = RunPCA(data, verbose = FALSE)
    plot(FeaturePlot(object = data_pca, features = c('nCount_RNA', 'percent.mt')))
    return(data_pca)
}
```


```{r}
run_umap = function(data, from_dimension, to_dimension) {
    data <- RunUMAP(data, dims = from_dimension:to_dimension, verbose = FALSE)

    data <- FindNeighbors(data, dims = from_dimension:to_dimension, verbose = FALSE)
    data <- FindClusters(data, verbose = FALSE)
    plot(DimPlot(data, label = TRUE) + NoLegend())
    return(data)
}
```



# Load datasets
```{r}
data_dir = file.path(dirname(getwd()), 'data')
print(data_dir)
```


```{r}
E13_hom = load_dataset(file.path(data_dir, 'E13_hom.csv'), 'E13_hom')
E14_hom = load_dataset(file.path(data_dir, 'E14_hom.csv'), 'E14_hom')
E13_het = load_dataset(file.path(data_dir, 'E13_het.csv'), 'E13_het')
E14_het = load_dataset(file.path(data_dir, 'E14_het.csv'), 'E14_het')
```


| Dataset | Genes | Cells |
|:------|:-----|:-----|
| E13_hom | 30,213 | 6,333 |
| E14_hom | 30,213 | 4,446 |
| E13_het | 30,213 | 3,797 |
| E14_het | 30,213 | 4,380 |



# Quality Control (QC)

We want to filter out cells based on technical or biological parameters. We explore QC metrics and filter out cells from each dataset, ideally being consistent for each technology (applying the same threshold for E13 hom/het and the same for E14 hom/het).

- E13 (hom/het): 10X Genomics v2.0
- E14 (hom/het): 10X Genomics v3.0

Below, we visualize gene counts (nFeature_RNA) and molecule counts (nCount_RNA), plot their relationship, and exclude cells with a clear outlier number of genes detected as potential multiplets. We also filter cells based on the percentage of mitochondrial genes present (percent.mt).

Plot distributions of QC metrics and select thresholds to filter out data.

Violin plot:

- **nFeature_RNA**: Number of unique genes detected in each cell: A few genes → low quality

- **nCount_RNA**: Total number of molecules detected within a cell (correlates with unique genes)

- **percent.mt**: Percentage of reads that map to the mitochondrial genome. High percentage of mitochondrial genes → low quality



```{r fig.width=13, fig.height=7}
E13_hom_qc = quality_control_plots(E13_hom, 'E13_hom')
```


```{r fig.width=13, fig.height=7}
E14_hom_qc = quality_control_plots(E14_hom, 'E14_hom')
```

```{r fig.width=13, fig.height=7}
E13_het_qc = quality_control_plots(E13_het, 'E13_het')
```

```{r fig.width=13, fig.height=7}
E14_het_qc = quality_control_plots(E14_het, 'E14_het')
```

## Remove cells based on a common threshold for each embryonic day

```{r}
E13_hom_subset = subset(E13_hom_qc, subset = (nFeature_RNA > 0 & nFeature_RNA < 6000) & (nCount_RNA > 1000 & nCount_RNA < 30000) & (percent.mt > 1 & percent.mt < 5))

E13_het_subset = subset(E13_het_qc, subset = (nFeature_RNA > 0 & nFeature_RNA < 6000) & (nCount_RNA > 1000 & nCount_RNA < 30000) & (percent.mt > 1 & percent.mt < 5))


E14_hom_subset = subset(E14_hom_qc, subset = (nFeature_RNA > 1000 & nFeature_RNA < 7000) & (nCount_RNA > 100 & nCount_RNA < 30000) & (percent.mt > 1 & percent.mt < 10))

E14_het_subset = subset(E14_het_qc, subset = (nFeature_RNA > 1000 & nFeature_RNA < 7000) & (nCount_RNA > 100 & nCount_RNA < 30000) & (percent.mt > 1 & percent.mt < 10))

```


| Dataset | Genes | Cells | Cells after QC |
|:------|:-----|:-----|:-----|
| E13_hom | 30,213 | 6,333 | 6,256 |
| E14_hom | 30,213 | 4,446 | 4,203 |
| E13_het | 30,213 | 3,797 | 3,717 |
| E14_het | 30,213 | 4,380 | 4,141 |


# SCTransform Normalisation

After removing unwanted cells from the datasets, we normalise the raw counts to account for differences in sequencing depth per cell for each sample by applying [sctransform normalization](https://satijalab.org/seurat/v3.0/sctransform_vignette.html). Sctransform models the UMI counts using a regularized negative binomial model to remove the variation due to sequencing depth (total nUMIs per cell), while adjusting the variance based on pooling information across genes with similar abundances (similar to some bulk RNA-seq methods).

Sctransform automatically regresses out sequencing depth (nUMIs); however, there are other sources of uninteresting variation in the data that is often specific to the dataset, such as the cell cycle phase as seen later.

SCTransform replaces *NormalizeData*, *ScaleData*, and *FindVariableFeatures*.

Transformed data will be available in the SCT assay, which is set as the default after running sctransform. We refer to each normalised dataset E13_hom_normalised etc as 'seurat_obj' to make the doc more abstract.

- `seurat_obj[["SCT"]]@scale.data` contains the residuals (normalized values), and is used directly as input to PCA. This matrix is non-sparse, and can therefore take up a lot of memory if stored for all genes. To save memory, we store these values only for variable genes, by setting the `return.only.var.genes = TRUE` by default in the SCTransform function call.
- The ‘corrected’ UMI counts are stored in `seurat_obj[["SCT"]]@counts`. We store log-normalized versions of these corrected counts in seurat_obj[["SCT"]]@data, which are very helpful for visualization.

During normalization, we can also remove confounding sources of variation, such as mitochondrial mapping percentage.

```{r}
E13_hom_normalised = SCTransform(E13_hom_subset, vars.to.regress = 'percent.mt', verbose = FALSE)
E14_hom_normalised = SCTransform(E14_hom_subset, vars.to.regress = 'percent.mt', verbose = FALSE)
E13_het_normalised = SCTransform(E13_het_subset, vars.to.regress = 'percent.mt', verbose = FALSE)
E14_het_normalised = SCTransform(E14_het_subset, vars.to.regress = 'percent.mt', verbose = FALSE)

```

Dataset dimensions after SCTransform:

| Dataset | Genes | Cells |
|:------|:-----|:-----|
| E13_hom | 15,356 | 6,256 |
| E14_hom | 15,785 | 4,203 |
| E13_het | 15,042 | 3,717 |
| E14_het | 15,648 | 4,141 |


# Dimensionality reduction

## Run PCA and plot first 2 principal components

```{r}
E13_hom_pca = run_plot_PCA(E13_hom_normalised)
```

```{r}
E14_hom_pca = run_plot_PCA(E14_hom_normalised)
```



```{r}
E13_het_pca = run_plot_PCA(E13_het_normalised)
```


```{r}
E14_het_pca = run_plot_PCA(E14_het_normalised)
```

## ElbowPlot to determine statistically significant PCs

ElbowPlot for E13_hom_pca. The plot is similar for the rest.

```{r}
ElbowPlot(E13_hom_pca, ndims = 50)
```



## Run UMAP on the reduced datasets and remove the Cajal-Retzius cells

[Cajal–Retzius cells](https://en.wikipedia.org/wiki/Cajal%E2%80%93Retzius_cell) (CR cells) (also known as Horizontal cells of Cajal) are a heterogeneous population of morphologically and molecularly distinct reelin-producing cell types in the marginal zone/layer I of the developmental cerebral cortex and in the immature hippocampus of different species and at different times during embryogenesis and postnatal life.


```{r fig.width=5, fig.height=5}
E13_hom_umap = run_umap(E13_hom_pca, 1, 30)
E13_hom_umap_subset = subset(E13_hom_umap, idents = c('10'), invert = TRUE)
plot(DimPlot(E13_hom_umap_subset, label = TRUE) + labs(title = 'E13_HOM'))
```


```{r fig.width=5, fig.height=5}
E14_hom_umap = run_umap(E14_hom_pca, 1, 30)
E14_hom_umap_subset = subset(E14_hom_umap, idents = c('13'), invert = TRUE)
plot(DimPlot(E14_hom_umap_subset, label = TRUE) + labs(title = 'E14_HOM'))
```


```{r fig.width=5, fig.height=5}
E13_het_umap = run_umap(E13_het_pca, 1, 30)
E13_het_umap_subset = subset(E13_het_umap, idents = c('11'), invert = TRUE)
plot(DimPlot(E13_het_umap_subset, label = TRUE) + labs(title = 'E13_HET'))
```


```{r fig.width=5, fig.height=5}
E14_het_umap = run_umap(E14_het_pca, 1, 30)
E14_het_umap_subset = subset(E14_het_umap, idents = c('13'), invert = TRUE)
plot(DimPlot(E14_het_umap_subset, label = TRUE) + labs(title = 'E14_HET'))
```




Final dataset dimensions:

| Dataset | Genes | Cells | Cells after filtering out |
|:------|:-----|:-----|:-----|
| E13_hom | 15,356 | 6,256 | 6,193 |
| E14_hom | 15,785 | 4,203 | 4,171 |
| E13_het | 15,042 | 3,717 | 3,653 |
| E14_het | 15,648 | 4,141 | 4,119 |



# Cell cycle scoring (to be included in the normalisation step in the next markdown file)

A known source of biological variation that can explain the data is the cell cycle.
Given a list of cell cycle phase marker genes in the textfile *regev_lab_cell_cycle_genes.txt*, we use the Seurat method [CellCycleScoring](https://satijalab.org/seurat/v3.1/cell_cycle_vignette.html) to assign scores to each cell for the cell cycle phases S and G2M. The cell is assigned to the phase with the highest score. If both scores are low, the cell is classified to the phase G1, even if it’s at the mitosis phase.

We perform cell cycle scoring on the full normalised dataset (before feature selection and dim. reduction)!

Cell cycle phases:

- **G1**: growth

- **S**: DNA Synthesis

- **G2M**: growth and preparation for mitosis


Load cell cycle genes from a regev publication. These are human genes so they need to be formatted to mouse genes. Then we separate them to S phase and G2M phase markers.

```{r}
human.cc <- scan('regev_lab_cell_cycle_genes.txt', what = 'character')
mouse.cc <- human.cc %>% tolower() %>% capitalize() %>% unlist
m.s <- mouse.cc[1:43]  # S: DNA synthesis
m.g2m <- mouse.cc[44:97]  # G2M: Growth and preparation for mitosis
```


CellCycleScoring stores S and G2/M scores in object meta data, along with the predicted classification of each cell in either G2M, S or G1 phase.

In the CellCycleScoring vignette attached above, CellCycleScoring is applied on normalised count data using `NormalizeData` while in the previous steps we used `SCTransform`, which combined normalisation, scaling and feature selection. Therefore, here we apply `NormalizeData` on the initial datasets after quality control, then run `CellCycleScoring`, which learns cell cycle scores that can be added to the `vars.to.regress` parameter in `SCTransform`. For all downstream analyses we can use the SCT assay.

```{r}
E13_hom_norm = NormalizeData(E13_hom_qc)
E13_hom_cell_cycle = CellCycleScoring(object = E13_hom_norm, s.features = m.s, g2m.features = m.g2m, set.ident = F)
E13_hom_norm = SCTransform(E13_hom_cell_cycle, vars.to.regress = 'percent.mt', verbose = FALSE)
```

